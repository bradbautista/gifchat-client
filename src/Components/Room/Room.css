form {
    background: lightgrey; 
    padding: 2rem;
    position: fixed; 
    bottom: 0; 
    display: flex;
    flex-direction: column;
    width: 98.5%;
}
  
form input { 
    border: 0; 
    padding: 10px;
    flex-grow: 1;
    margin-right: 2rem;
}
  
form button { 
    background: rgb(130, 224, 255); 
    border: none; 
    padding: 10px;
}
  
#messages { 
    list-style-type: none;
    margin: 0; 
    padding: 0; 
}
  
#messages li { 
    padding: 2rem; 
}

#messages img { 
    border-radius: 1.5rem; 
}
  
#messages li:nth-child(odd) { 
    background: #eee; 
}

.hidden {
    display: none;
}

.unhidden {
    display: flex;
}

.input-flex-wrapper {
    display: flex;
    flex-direction: row;
    flex-grow: 1;
}

#gif-options {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: flex-start;
    padding: 0;
    margin: 0;
    margin-top: 2rem;
}

#gif-options img {
    margin: 1rem;
    border-radius: 1.5rem;
}