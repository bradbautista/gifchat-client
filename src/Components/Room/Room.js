import React, { Component } from 'react';
import './Room.css';
import io from 'socket.io-client';
import { render } from '@testing-library/react';
import config from '../../config'
const uuidv4 = require('uuid/v4');

export default class Room extends Component {

  constructor(props) {
    super(props);
    this.state = {
      value: '',
      messages: [],
      gifs: {
        previews: [],
        fullsize: [],
      }
    };

  }

  
  sendMessage = (msg) => {
  
    console.log(this.state.messages)
    console.log(msg)

    // HI BRAD THIS IS BRAD THE CODE BELOW IS JUST FOR THE DUMMY CLIENT, THE CODE BELOW THAT IS GOOD AND YOU SHOULD UNCOMMENT IT AT THE PROPER TIME BUT FOR NOW WE'RE JUST PUSHING THE IMAGE INTO THE MESSAGES ARRAY

    this.setState({ gifs: {previews: [], fullsize: []} })
    this.setState({ value: '' })
    this.setState({messages: [...this.state.messages, msg]})

    // Designate the server socket to emit to; this will have
    // to be the room name, i.e. url (path?); when we get there, we'll 
    // need to install and require unique names, but for now it's just
    // going to connect to the proof-of-concept server
    // const socket = io('http://localhost:3000/');

    // Emit a chat message to the server; rather than
    // this.state.value, the value after the function
    // name changes will be the GIF url. Don't know
    // whether we'll need a setState statement after
    // socket.emit('chat message', this.state.value);
    //       this.setState({value: ''})
  
  }

  // Query the Tenor API for GIFs related to the search term, then set those GIFs in state as GIF options

  getGifs = (e) => {
    
    e.preventDefault()

    // Clear the previous options
    this.setState({previews: []})

    const endpoint = config.API_ENDPOINT
    const query = this.state.value
    const apiKey = config.API_KEY
    const limit = 8 // The number of gifs to fetch

    const url = `${endpoint}search?q=${query}&key=${apiKey}&limit=${limit}`
    const options = {
      method: 'GET',
      redirect: 'follow',
    }

    // console.log(url)
    // console.log(endpoint)
    // console.log(query)
    // console.log(this.state)

    return fetch(url, options)
        .then((res) => {
          return res.json();
        })
        .then((responseJson) => {

          // Trim the results
          const searchResults = responseJson.results

          // Extract previews from the results
          const gifPreviews = searchResults.map((result) => {
            return result.media[0].tinygif.url
          })

          // Extract the full-size gifs from the results so we can send the full-size gif to chat
          const fullSizeGifs = searchResults.map((result) => {
            return result.media[0].gif.url
          })
          
          // Set the gifs
          this.setState({
            gifs: {
              previews: [...gifPreviews], 
              fullsize: [...fullSizeGifs]
            }
          })
          
        })
        .catch(error => { console.error(error) })
  
  }

  // Updates state with search term as user types
  handleChange = (e) => {
    
    this.setState({value: e.target.value});
  
  }

  emptyOptions = (e) => {
    
    let n = Math.floor(Math.random() * Math.floor(7));

    const prngArray = this.state.gifs.previews.splice(0, n);

    console.log(prngArray)

    e.preventDefault()
    this.setState({gifs: {previews: [...prngArray]}});
    // console.log('Heya')
  
  }

  // When the server emits a msg to the client,
  // update this.state.messages to include the msg
  handleMessage = (msg) => {

    console.log(msg)
    
    this.setState({messages: [...this.state.messages, msg]})

    console.log(this.state.messages)

  }

  // When the page loads, 
  componentDidMount() {

    // console.log(this)

    // console.log(this.state.messages)

    // Establish which socket to communicate with;
    // this will ultimately be the room URL, but for
    // now it's just the dummy server
    // const socket = io('http://localhost:3000');

    // When the 
    // socket.on('chat message', this.handleMessage)

  }
  

  render() {

    // console.log(this.state.value)

    const msgs = this.state.messages.map((msg) => {
        return (
            <li key={uuidv4()}>
                <img src={msg}></img>
            </li>
        )
    })

    const gifOptions = this.state.gifs.previews.map((preview, i) => {

        // We don't want to send the preview to the server, we want to send the full-size gif

        const fullSizeGif = this.state.gifs.fullsize[i]

        // Select small preview gif and arrange them in a flex container that wraps
        return (
            <li key={i}>
                <img onClick={() => this.sendMessage(fullSizeGif)} className="chat-message" src={preview} />
            </li>
        )
    })

  return (
    <div className="Room">
    <main>

        <ul className="messages">
        {msgs}
        </ul>

        <form onSubmit={this.getGifs}>
        {/* <form onSubmit={this.emptyOptions}> */}

          <div className="room-input-flex-wrapper">
            <input type="text" 
              value={this.state.value} 
              onChange={this.handleChange} 
              id="m" 
              autoComplete="off" 
              placeholder="Enter search term"
            />
            <button>Search</button>
          </div>

          <img className="tenor-attr" src="https://www.gstatic.com/tenor/web/attribution/PB_tenor_logo_blue_horizontal.png" />

          {/* React wants to re-render the form component every time the array (and thus height) changes, which kills our CSS transition. To get around this, we generate a dynamic height value based on the length of the array. */}

          <ul className="gif-options" 
          style={{ height: this.state.gifs.previews.length * 15 + 'rem'}}
          >
          {gifOptions}
          </ul>

        </form>
    </main>

    </div>
  );

  }
}

