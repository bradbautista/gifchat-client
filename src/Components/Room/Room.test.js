import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter } from 'react-router-dom';
import Room from './Room';
import renderer from 'react-test-renderer';

// Router wraps everything in index.js,
// but we need to wrap it here else the test complains
// about not having one

it('Renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<BrowserRouter><Room /></BrowserRouter>, div);
  ReactDOM.unmountComponentAtNode(div);
});


it('Renders the UI expected', () => {
    let error = false;
    const someText = 'A search term'
    const someFunc = () => {};
    const msgs = [
        'https://www.someurl.com/gif.gif', 
        'https://www.someurl.com/gif.gif', 
        'https://www.someurl.com/gif.gif'
    ];
    const gifOptions = [
        'https://www.someurl.com/gif.gif', 
        'https://www.someurl.com/gif.gif', 
        'https://www.someurl.com/gif.gif'
    ];
    const tree = renderer
        .create(
            <BrowserRouter>
                <ul className="messages">
                    <div
                    style={{ display: error ? '' : 'none' }}
                    className="error"
                    >
                    Error: Messages could not be displayed. The room does not exist,
                    is full, or the Internet is broken. Try refreshing, or
                    <a href="https://gifchat.now.sh/" style={{ fontWeight: 700 }}>
                        {' '}
                        leave this place.
                    </a>
                    </div>
                    {msgs.reverse()}
                </ul>

                <form onSubmit={someFunc}>
                    <div className="room-input-flex-wrapper">
                    <input
                        type="text"
                        value={someText}
                        onChange={someFunc}
                        id="m"
                        autoComplete="off"
                        placeholder="Enter search term"
                    />
                    <button>Search</button>
                    </div>

                    <img
                    alt=""
                    className="tenor-attr"
                    src="https://www.gstatic.com/tenor/web/attribution/PB_tenor_logo_blue_horizontal.png"
                    />

                    {/* React wants to re-render the form component every time the array (and thus height) changes, which kills our CSS transition. To get around this, we generate a dynamic height value based on the length of the array. */}
                    
                    <ul
                    className="gif-options"
                    style={{ height: gifOptions.length * 15 + 'rem' }}
                    >
                    {gifOptions}
                    <button 
                    className='get-gifs-btn' 
                    // Do not display the button if:
                    // - There are fewer than 10 results for your search, or
                    // - You've run out of additional search results
                    style={{ 
                        display: (msgs.length < 10) ? 'none' : (msgs.length > 10 && msgs.length === '0') ? 'none' : ''}}
                    onClick={someFunc}
                    >MORE GIFS!</button>
                    </ul>
                </form>
            </BrowserRouter>
            )
        .toJSON();
    expect(tree).toMatchSnapshot();
        
})