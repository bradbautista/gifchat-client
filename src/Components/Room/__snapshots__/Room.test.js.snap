// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the UI expected 1`] = `
Array [
  <ul
    className="messages"
  >
    <div
      className="error"
      style={
        Object {
          "display": "none",
        }
      }
    >
      Error: Messages could not be displayed. The room does not exist, is full, or the Internet is broken. Try refreshing, or
      <a
        href="https://gifchat.now.sh/"
        style={
          Object {
            "fontWeight": 700,
          }
        }
      >
         
        leave this place.
      </a>
    </div>
    https://www.someurl.com/gif.gif
    https://www.someurl.com/gif.gif
    https://www.someurl.com/gif.gif
  </ul>,
  <form
    onSubmit={[Function]}
  >
    <div
      className="room-input-flex-wrapper"
    >
      <input
        autoComplete="off"
        id="m"
        onChange={[Function]}
        placeholder="Enter search term"
        type="text"
        value="A search term"
      />
      <button>
        Search
      </button>
    </div>
    <img
      alt=""
      className="tenor-attr"
      src="https://www.gstatic.com/tenor/web/attribution/PB_tenor_logo_blue_horizontal.png"
    />
    <ul
      className="gif-options"
      style={
        Object {
          "height": "45rem",
        }
      }
    >
      https://www.someurl.com/gif.gif
      https://www.someurl.com/gif.gif
      https://www.someurl.com/gif.gif
      <button
        className="get-gifs-btn"
        onClick={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
      >
        MORE GIFS!
      </button>
    </ul>
  </form>,
]
`;
